source/Patches/EndGameSummary.cs

```diff
@@ -33,14 +39,49 @@ internal class PlayerInfo
             }
         }
 
+        [HarmonyPatch(typeof(HudManager), nameof(HudManager.Update))]
+        [HarmonyPriority(Priority.Last)]
+        public static void Postfix(HudManager __instance)
+        {
+            if (EndGameSummary.AdditionalTempData.updateFlag)
+            {
+                //EndGameSummary.TrackRoleHistory();
+                EndGameSummary.UpdatePlayerInfo();
+            }
+        }
+
+        public static void TrackRoleHistory()
+        {
+
+            //this might work better if disconnections seem to be a problem.. player.name would need changing to player.PlayerName though.
+            //foreach (var player in GameData.Instance.AllPlayers)
+            foreach (var player in PlayerControl.AllPlayerControls)
+            {
+                var role = Role.GetRole(player);
+                if (role != null && (AdditionalTempData.roleHistory.FindAll(x => x.Key == player.PlayerId).Count==0 ||AdditionalTempData.roleHistory.Last(x => x.Key == player.PlayerId).Value != role))
```

Comment:
This find is a bit cumbersome; I wonder if instead of `List<KeyValuePair<byte, Role>>`, something like `Map<byte, List<Role>>` might work better?

Reply:
I fiddled around a lot with RoleEnum vs Role, i can have another fiddle at another time, but my time for coding atm is really limited, it works as is.

regarding storing the role history inside player info, its a lot harder to access rolehistory within playerinfo, i tried to integrate it in and found it much simpler to keep it seperate.

regarding Map<byte,List<Role>>, i am not aware of "Map" or what it does, the aim of the list of KeyValuePair's is to then have a list of player IDs to cycle through, to which you can pull the roles in order of addition to the list, there maybe a better way of doing it, but i was not aware at the time of writing it.

Reply:
C# might call it a Dictionary instead of a Map (well, a Map is a kind of associative type and a Dictionary is a Map implemented with a hashtable). Constant time accesses instead of linear.

Reply:
If I get time to look at this I will look into changing it.

Reply:
ok, had a look at it, and the reason why they are seperate is because the game information is collected from scratch every time it runs, it ignores the previous information, and re-aquires it.

so to change this i would have to change how it all works, so instead of it always adding new data, it replaces old data, this will take some time